python/shotgun_model/data_handler.py

```diff
@@ -0,0 +1,473 @@
+            self.CACHE_BY_UID: {},    # uid-based lookup
+            self.UID: None            # the uid of the root is None
+        }
+
+    def _clear_cache(self):
+        """
+        Sets up an empty cache in memory
+        """
+        self._cache = self._create_clear_cache()
+
+    def is_cache_available(self):
+        """
+        Returns true if the cache exists on disk, false if not.
+
+        :returns: boolean to indicate if cache exists on disk
+        """
+        return os.path.exists(self._cache_path)
+
+    def is_cache_loaded(self):
+        """
+        Returns true if the cache has been loaded into memory, false if not.
+
+        :returns: boolean to indicate if cache is loaded
+        """
+        return self._cache is not None
+
+    @log_timing
+    def remove_cache(self):
+        """
+        Removes the associated cache file from disk
+        and unloads cache data from memory.
+
+        :returns: True if the cache was sucessfully unloaded.
+        """
+        if os.path.exists(self._cache_path):
+            try:
+                os.remove(self._cache_path)
+            except Exception, e:
+                self._log_warning(
+                    "Could not remove cache file '%s' "
+                    "from disk. Details: %s" % (self._cache_path, e)
+                )
+                return False
+        else:
+            self._log_debug("...no cache file found on disk. Nothing to remove.")
+
+        # unload from memory
+        self.unload_cache()
+
+        # init again
+        self.load_cache()
+
+        return True
+
+    @log_timing
+    def load_cache(self):
+        """
+        Loads a cache from disk into memory
+        """
+        # init empty cache
+        self._clear_cache()
+
+        # try to load
+        self._log_debug("Loading from disk: %s" % self._cache_path)
+        if os.path.exists(self._cache_path):
+            try:
+                with open(self._cache_path, "rb") as fh:
+                    pickler = cPickle.Unpickler(fh)
+                    file_version = pickler.load()
+                    if file_version != self.FORMAT_VERSION:
+                        raise ShotgunModelDataError(
+                            "Cache file has version %s - version %s is required" % (file_version, self.FORMAT_VERSION)
+                        )
+                    self._cache = pickler.load()
+            except Exception, e:
+                self._log_debug("Cache '%s' not valid - ignoring. Details: %s" % (self._cache_path, e))
+
+        else:
+            self._log_debug("No cache found on disk. Starting from empty data store.")
+
+        self._log_debug("Cache load complete: %s" % self)
+
+    def unload_cache(self):
+        """
+        Unloads any in-memory cache data.
+        """
+        if self._cache is None:
+            # nothing to do
+            return
+
+        self._log_debug("Unloading in-memory cache for %s" % self)
+        self._cache = None
```

Comment:
i think i'm with JF on this one. `None` for me indicates that something is un-initialized or not present. The data handler has two states, loaded or unloaded, and in the unloaded state the cache is set to None. The methods load, unload and save (and remove) control the state of the data handler. I agree that the data structure is a little confusing, it's intentionally raw so that it is straight forward to pickle out to disk.